### PR TITLE
Stylo: Basic support for grid-{row,column}-{start,end}

### PR DESCRIPTION
**Pull request by ghost**
---

- [x] `./mach build-geckolib` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [ ] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14000)

### CODE REVIEW

**Comment by Wafflespeanut** [comment]
@Manishearth (cc @emilio) This ain't finished yet! It's a rough setup for some of the grid-* short and longhands. I'm not quite sure whether you people are convinced with the `GridLine` type (just need your comments, don't review the whole thing!).

File: components/style/properties/shorthand/position.mako.rs [hunk 1]
```diff
@@ -119,3 +121,85 @@
...
+        let mut row_start = Default::default();
+        let mut column_start = Default::default();
+        let mut row_end = Default::default();
+        let mut column_end = Default::default();
+
+        if let Ok(line) = input.try(GridLine::parse) {
+            row_start = line;
+        }
+
+        if let Ok(_) = input.try(|i| i.expect_delim('/')) {
```
Comment (emilio) [inline components/style/properties/shorthand/position.mako.rs]: nit: `if input.try(..).is_ok()`, here and below.

File: components/style/properties/shorthand/position.mako.rs [hunk 2]
```diff
@@ -119,3 +121,85 @@
         }
     }
 </%helpers:shorthand>
+
+<%helpers:shorthand name="grid-area" products="gecko"
+                    sub_properties="grid-row-start grid-column-start grid-row-end grid-column-end">
+    use values::specified::grid::GridLine;
+
+    pub fn parse_value(_: &ParserContext, input: &mut Parser) -> Result<Longhands, ()> {
+        let mut row_start = Default::default();
```
Comment (emilio) [inline components/style/properties/shorthand/position.mako.rs]: Please use `GridLine::default` so it's clearer what's the type of these.

File: components/style/properties/shorthand/position.mako.rs [hunk 3]
```diff
@@ -119,3 +121,85 @@
         }
     }
 </%helpers:shorthand>
+
+<%helpers:shorthand name="grid-area" products="gecko"
```
Comment (emilio) [inline components/style/properties/shorthand/position.mako.rs]: Spec link?

File: components/style/properties/longhand/position.mako.rs [hunk 4]
```diff
@@ -181,3 +182,29 @@ ${helpers.single_keyword("box-sizing",
...
+    }
+
+    pub type SpecifiedValue = GridLine;
+
+    #[inline]
+    pub fn get_initial_value() -> computed_value::T {
+        Default::default()
+    }
+
+    pub fn parse(_: &ParserContext, input: &mut Parser) -> Result<SpecifiedValue, ()> {
```
Comment (emilio) [inline components/style/properties/longhand/position.mako.rs]: `#[inline]`?

File: components/style/values/specified/grid.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        let mut grid_line = Default::default();
+        if let Ok(_) = input.try(|i| i.expect_ident_matching("auto")) {
+            return Ok(grid_line)
+        }
+
+        for _ in 0..3 {     // Maximum possible entities for <grid-line>
+            if let Ok(_) = input.try(|i| i.expect_ident_matching("span")) {
+                if grid_line.is_span {
+                    return Err(())
+                } else {
```
Comment (emilio) [inline components/style/values/specified/grid.rs]: no need for `else` blocks after a `return`, here and below.

File: components/style/values/specified/grid.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        if let Some(i) = self.integer {
+            vec.push(format!("{}", i));
+        }
+
+        if let Some(ref s) = self.ident {
+            vec.push(s.to_owned());
+        }
+
+        write!(dest, "{}", vec.join(" "))
```
Comment (emilio) [inline components/style/values/specified/grid.rs]: Seems like using a vector and allocating for this is overkill, and a single boolean variable should do to get the same effect.

**Review by emilio — CHANGES_REQUESTED** [review]
I still need to review the actual parsing logic, this is only a superficial review.

**Comment by emilio** [comment]
Oh, and the `GridLine` type sounds good to me, and aligns nicely with what Gecko does in this case, so...

**Comment by asajeffrey** [comment]
@Wafflespeanut is this still active?

**Comment by Wafflespeanut** [comment]
Yes, sorry about that. It's still active :smile:

**Comment by asajeffrey** [comment]
Glad to hear it!

File: components/style/properties/longhand/position.mako.rs [hunk 7]
```diff
@@ -181,3 +182,31 @@ ${helpers.single_keyword("box-sizing",
 // https://drafts.csswg.org/css-images-3/
 ${helpers.single_keyword("object-fit", "fill contain cover none scale-down",
                          products="gecko", animatable=False)}
+
+// https://drafts.csswg.org/css-grid/#propdef-grid-row-start
+<% grid_longhands = ["grid-row-start", "grid-row-end", "grid-column-start", "grid-column-end"] %>
+
+% for longhand in grid_longhands:
+<%helpers:longhand name="${longhand}" animatable="False" products="gecko">
```
Comment (Manishearth) [inline components/style/properties/longhand/position.mako.rs]: Can we use `predefined_type` here?

File: components/style/properties/gecko.mako.rs [hunk 8]
```diff
@@ -738,6 +745,24 @@ fn static_assert() {
     }
     ${impl_simple_copy('box_sizing', 'mBoxSizing')}
 
+    % for value in GRID_LINES:
+    pub fn set_${value.name}(&mut self, v: longhands::${value.name}::computed_value::T) {
```
Comment (Manishearth) [inline components/style/properties/gecko.mako.rs]: http://searchfox.org/mozilla-central/source/layout/style/nsRuleNode.cpp#8314 is what Gecko does. I see a couple of things missing:

 - The auto unit isn't handled here. Not sure if it needs to be; but there should be comments about how this is represented in gecko
 - Clamping between the maximum allowed unit isn't handled here.

File: components/style/values/specified/grid.rs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use cssparser::{Parser, ToCss};
+use std::fmt;
+use values::NoViewportPercentage;
+use values::computed::ComputedValueAsSpecified;
+
+#[derive(PartialEq, Clone, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct GridLine {
```
Comment (Manishearth) [inline components/style/values/specified/grid.rs]: spec link to https://drafts.csswg.org/css-grid/#typedef-grid-row-start-grid-line please 😄

File: components/style/properties/shorthand/position.mako.rs [hunk 10]
```diff
@@ -119,3 +121,87 @@
...
+            grid_column_start: Some(column_start),
+            grid_row_end: Some(row_end),
+            grid_column_end: Some(column_end),
+        })
+    }
+
+    impl<'a> LonghandsToSerialize<'a>  {
+        fn to_css_declared<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            try!(self.grid_row_start.to_css(dest));
+            try!(self.grid_column_start.to_css(dest));
```
Comment (Manishearth) [inline components/style/properties/shorthand/position.mako.rs]: Shouldn't there be slashes here?

File: components/style/properties/shorthand/position.mako.rs [hunk 11]
```diff
@@ -119,3 +121,87 @@
...
+// https://drafts.csswg.org/css-grid/#propdef-grid-row
+% for prop in ["grid-row", "grid-column"]:
+<%helpers:shorthand name="${prop}" products="gecko" sub_properties="${prop}-start ${prop}-end">
+    use values::specified::grid::GridLine;
+
+    pub fn parse_value(_: &ParserContext, input: &mut Parser) -> Result<Longhands, ()> {
+        let mut start = GridLine::default();
+        let mut end = GridLine::default();
+
+        if let Ok(line) = input.try(GridLine::parse) {
```
Comment (Manishearth) [inline components/style/properties/shorthand/position.mako.rs]: This should be a `try!`; if this parse fails then there is no property.

**Review by Manishearth — APPROVED** [review]
Minor nits, so far the properties look good. Is this enough to get something working in Stylo? Or are there other properties that need to be implemented?

**Comment by Wafflespeanut** [comment]
I'm planning to land the `GridLine` type for now (after testing of course). I've left the shorthand parsing/serialization for the newcomers.

**Comment by Manishearth** [comment]
What's the status of this?

**Comment by Wafflespeanut** [comment]
Silly me. I keep forgetting about this. Sorry for keeping this so long. And no, it doesn't need anything else. It works!

screenshot

File: components/style/properties/longhand/position.mako.rs (hunk 7 above)
Comment (Wafflespeanut) [inline components/style/properties/longhand/position.mako.rs]: Yeah, thanks for reminding :smile:

File: components/style/properties/gecko.mako.rs (hunk 8 above)
Comment (Wafflespeanut) [inline components/style/properties/gecko.mako.rs]: - [Empty string, `0i32` and no span](https://dxr.mozilla.org/mozilla-central/rev/6dbc6e9f62a705d5f523cc750811bd01c8275ec6/layout/style/nsStyleStruct.h#1608-1621) should indicate `auto`, right?
- Clamping done :)

File: components/style/properties/gecko.mako.rs [hunk 12]
```diff
@@ -608,8 +608,15 @@ class Corner(object):
         self.x_index = 2 * index
         self.y_index = 2 * index + 1
 
+class GridLine(object):
```
Comment (Manishearth) [inline components/style/properties/gecko.mako.rs]: It's used very few times basically. I guess it's okay.

**Pull request merged**